torch/_dynamo/side_effects.py

```diff
@@ -15,53 +14,42 @@
 from .exc import unimplemented
 from .source import LocalSource, Source
 from .utils import nn_module_new, object_new
-from .variables.base import VariableTracker
+from .variables.base import is_side_effect_safe, MutableLocalBase, VariableTracker
 
 
-@dataclasses.dataclass
-class MutableSideEffects:
+class MutableSideEffects(MutableLocalBase):
     """
     VariableTracker.mutable_local marker to indicate a list passed as
     an input that if we mutate we need to re-apply those mutations after
     the graph runs.
     """
 
-    source: Source
-    is_modified: bool = False
+    def __init__(self, source: Source, is_modified: bool = False):
+        super().__init__("existing")
+        self.source = source
+        self.is_modified = is_modified
 
-    def __hash__(self):
-        return id(self)
 
-    def __eq__(self, other):
-        return self is other
-
-
-@dataclasses.dataclass
-class AttributeMutation:
+class AttributeMutation(MutableLocalBase):
     """
     VariableTracker.mutable_local marker to track changes to attributes
     """
 
-    source: Source
+    def __init__(self, typ: str, source: Source):
+        super().__init__(typ)
+        self.source = source
 
 
 class AttributeMutationExisting(AttributeMutation):
-    def __hash__(self):
-        return id(self)
-
-    def __eq__(self, other):
-        return self is other
+    def __init__(self, source: Source):
+        super().__init__("existing", source)
+        self.source = source
 
 
-@dataclasses.dataclass
```

Comment:
Where have the dataclasses gone? have you run afoul of some gross dataclass inheritance/default value / property warts?

Reply:
We now need to specify whether a mutable_local is from an existing variable that Dynamo begins to track or if the mutable local is a new variable created inside the function Dynamo is introspecting.

We use this information to assign a scope id.

This information gets specified via the constructor to MutableLocalBase, which we need to get at by actually writing constructors for these classes, so I made these regular classes so we can specify constructors for them.